service/history/worker_versioning_util.go

```diff
@@ -0,0 +1,167 @@
+package history
+
+import (
+	"context"
+
+	persistencespb "go.temporal.io/server/api/persistence/v1"
+	taskqueuespb "go.temporal.io/server/api/taskqueue/v1"
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/worker_versioning"
+	"go.temporal.io/server/service/history/consts"
+	"go.temporal.io/server/service/history/shard"
+	"go.temporal.io/server/service/history/tasks"
+	"go.temporal.io/server/service/history/workflow"
+	"go.temporal.io/server/service/history/workflow/cache"
+)
+
+func updateIndependentActivityBuildId(
+	ctx context.Context,
+	task tasks.Task,
+	scheduledEventId int64,
+	buildId string,
+	shardContext shard.Context,
+	workflowCache cache.Cache,
+	metricsHandler metrics.Handler,
+	logger log.Logger,
+) (retErr error) {
+	if buildId == "" {
+		// the task is sync-matched, or versioning is not enabled for this Task Queue
+		return nil
+	}
+
+	defer func() {
+		if retErr != nil {
+			// An independent task is scheduled but we could not persist the assigned build ID in mutable state.
+			// This is only an issue for user visibility: until the task is started user would not see the assigned
+			// build ID.
+			// Since it does not affect WF progress, just logging warn and skipping the error.
+			// TODO: let the error bubble up so the task is rescheduled and build ID is fully updated
+			logger.Warn("failed to update activity's assigned build ID", tag.Error(retErr))
+		}
+		retErr = nil
+	}()
+
+	weContext, release, err := getWorkflowExecutionContextForTask(ctx, shardContext, workflowCache, task)
```

Comment:
can you add some comments about how often we get to this point and have to do another update?

Reply:
added a comment for it. TLDR; these updates should be relatively rare.